aws-predictions-tfl/src/main/java/com/amplifyframework/predictions/service/TFLiteTextClassificationService.java

```diff
@@ -0,0 +1,332 @@
+     * @param onError notified when classification fails
+     */
+    void classify(
+            @NonNull String text,
+            @NonNull Consumer<InterpretResult> onSuccess,
+            @NonNull Consumer<PredictionsException> onError
+    ) {
+        final Sentiment sentiment;
+        try {
+            loadIfNotLoaded();
+            sentiment = fetchSentiment(text);
+        } catch (PredictionsException exception) {
+            onError.accept(exception);
+            return;
+        }
+
+        InterpretResult result = InterpretResult.builder()
+                .sentiment(sentiment)
+                .build();
+        onSuccess.accept(result);
+    }
+
+    private Sentiment fetchSentiment(String text) throws PredictionsException {
+        float[][] input;
+        float[][] output;
+
+        // Pre-process input text
+        try {
+            input = tokenizeInputText(text);
+            output = new float[1][labels.size()];
+        } catch (IllegalArgumentException exception) {
+            throw new PredictionsException(
+                    "Tensorflow Lite failed to make inference.",
+                    exception,
+                    "Verify that the assets are loaded."
+            );
+        }
+
+        // Run inference.
+        tflite.run(input, output);
+
+        // Find the predominant sentiment
+        Sentiment sentiment = null;
+        for (int i = 0; i < labels.size(); i++) {
+            SentimentType sentimentType = SentimentTypeAdapter.fromTensorflow(labels.get(i));
+            float confidenceScore = output[0][i] * PERCENT;
+            if (sentiment == null || sentiment.getConfidence() < confidenceScore) {
+                sentiment = Sentiment.builder()
+                        .value(sentimentType)
+                        .confidence(confidenceScore)
+                        .build();
+            }
+        }
+        return sentiment;
+    }
+
+    @SuppressWarnings("ConstantConditions")
+    private float[][] tokenizeInputText(String text) {
+        float[] tmp = new float[SENTENCE_LEN];
+
+        int index = 0;
+        tmp[index++] = dictionary.get(START);
+
+        for (String word : text.split(SIMPLE_SPACE_OR_PUNCTUATION)) {
+            if (index >= SENTENCE_LEN) {
+                break;
+            }
+
+            tmp[index++] = dictionary.containsKey(word)
+                    ? dictionary.get(word)
+                    : dictionary.get(UNKNOWN);
+        }
+
+        // Padding and wrapping.
+        Arrays.fill(tmp, index, SENTENCE_LEN - 1, dictionary.get(PAD));
+        return new float[][]{tmp};
+    }
+
+    @WorkerThread
+    private synchronized void loadIfNotLoaded() throws PredictionsException {
+        if (loaded.get()) {
+            return;
+        }
+        loadModel();
+        loadDictionary();
+        loadLabels();
+        loaded.set(true);
+    }
+
+    @WorkerThread
+    private synchronized void loadModel() throws PredictionsException {
```

Comment:
You have a number of different components that have a similar lifecycle: `load()`, `isLoaded()`.

You could make a `Loadable` interface:
```
// V = value type for thing that is loaded, e.g. a Map<String, Integer> dictionary, etc.
interface Loadable<V> {
    void load();
    void unload();
    boolean isLoaded();
    onLoaded(Consumer<V> valueConsumer);
    onUnloaded(Action unloadAction);
    @Nullable V value()
}
```

If you have somehting like this, you can externalize all of these details into small little template files and compartmentalize the load/validation logic into them.

Then the classification refers to them more like:
```
for (Loadable<?> dependency : LOADABLES) {
    if (!dependency.isLoaded()) {
        dependency.load();
    }
}
```

I guess you'd want this to happen as a down-stream result of the call-chain from the`Plugin<?>`'s `initialize()` lifecycle method.

Reply:
addressed